### PR TITLE
egressgw: consolidate and reoptimise EgressRules

### PR DESCRIPTION
**Pull request by markpash**
> egressgw: consolidate and reoptimise EgressRules
> 
> The addMissingEgressRules and removeUnusedEgressRules methods are quite
> confusing to read. Consolidate these into one method.
> 
> Delete the policyMatches method and the sourceIP optimisation. Given
> that all necessary egress rules are generated in the upsert part of the
> bpf map reconciliation process, we can just dump them all into a map and
> do simple lookups.

PR best viewed in isolation, the GitHub diff is very confusing.

Justification for changes:
- One less item of state kept in the Manager (`policyConfigsBySourceIP`)
- Previous code was difficult to read
- Reduce passed around variables and hard to follow code paths
- Relatively clean inlining and removing single-use or redundant methods
- I'm pretty confident this is generally more efficient
- The resulting effect should be the same (map should contain the same contents)

### CODE REVIEW

File: pkg/maps/egressmap/policy.go [hunk 1]
```diff
@@ -178,14 +177,12 @@ func (k *EgressPolicyKey4) Match(sourceIP netip.Addr, destCIDR netip.Prefix) boo
 
 // GetSourceIP returns the egress policy key's source IP.
 func (k *EgressPolicyKey4) GetSourceIP() netip.Addr {
-	addr, _ := netipx.FromStdIP(k.SourceIP.IP())
-	return addr
+	return k.SourceIP.Addr()
```
Comment (ysksuzuki) [inline pkg/maps/egressmap/policy.go:180]: Just to be clear, `FromStdIP` implicitly unmaps IPv6-mapped IPv4 addresses, but we don't need it because we don't accept such addresses here, right?

File: pkg/egressgateway/manager_privileged_test.go [hunk 2]
```diff
@@ -634,3 +635,110 @@ func tryAssertEgressRules(policyMap egressmap.PolicyMap, rules []egressRule) err
...
+}
+
+func (fm *FakeEgressPolicyMap) IterateWithCallback(cb egressmap.EgressPolicyIterateCallback) error {
+	for k, v := range fm.backingMap {
+		cb(&k, &v)
+	}
+	return nil
+}
+
+func BenchmarkReconcileMap(b *testing.B) {
```
Comment (ysksuzuki) [inline pkg/egressgateway/manager_privileged_test.go:675]: How do we know if there's performance regression happens with this test? Does it fail if some threshold is exceeded?

File: pkg/egressgateway/manager.go [hunk 3]
```diff
@@ -543,141 +538,96 @@ func (manager *Manager) updatePoliciesMatchedEndpointIDs() {
...
-		matchPolicy := func(endpointIP netip.Addr, dstCIDR netip.Prefix, excludedCIDR bool, gwc *gatewayConfig) bool {
-			gatewayIP := gwc.gatewayIP
-			if excludedCIDR {
-				gatewayIP = ExcludedCIDRIPv4
-			}
-
-			return policyKey.Match(endpointIP, dstCIDR) && policyVal.Match(gwc.egressIP, gatewayIP)
+	// Remove stuff that's not present in the desired state.
+	for key, val := range bpfEgressState {
+		if _, exists := desired[key]; exists {
```
Comment (ysksuzuki) [inline pkg/egressgateway/manager.go:612]: I think the original logic in `removeUnusedEgressRules` checks if both policy key and value matches the desired state. Is it OK not to do the same here?

File: pkg/egressgateway/manager.go [hunk 4]
```diff
@@ -543,141 +538,96 @@ func (manager *Manager) updatePoliciesMatchedEndpointIDs() {
...
-				isExcludedCIDR := false
-				for _, dstCIDR := range policy.dstCIDRs {
-					if f(endpointIP, dstCIDR, isExcludedCIDR, &policy.gatewayConfig) {
-						return true
-					}
+// reconcileEgressRules takes all of the current manager state and
+// safely reconciles with the bpf map.
+func (manager *Manager) reconcileEgressRules() {
+	// Generate our desired state.
+	desired := map[egressmap.EgressPolicyKey4]egressmap.EgressPolicyVal4{}
```
Comment (ysksuzuki) [inline pkg/egressgateway/manager.go:551]: Now we have 2 `map[egressmap.EgressPolicyKey4]egressmap.EgressPolicyVal4{}`, desired and actual. We don't have to be worried about the memory overhead that may cause?

**Review by ysksuzuki — CHANGES_REQUESTED** [review]

File: pkg/egressgateway/manager_privileged_test.go (hunk 2 above)
Comment (markpash) [reply pkg/egressgateway/manager_privileged_test.go:675]: I've included this to be able to add some numbers to support the change. I don't believe we have performance regression testing at the moment.

File: pkg/egressgateway/manager.go (hunk 3 above)
Comment (markpash) [reply pkg/egressgateway/manager.go:612]: I believe so. The code slightly further up from here will Add and also update existing keys, so all keys that should exist in the datapath will be present by this point in the code. (and also contain the correct value)

The highlighted part of the code simply makes sure that any keys that are present in the datapath that aren't desired, get deleted. For deletion, we don't care what value that held, we already know that the key must be deleted.

File: pkg/egressgateway/manager.go (hunk 4 above)
Comment (markpash) [reply pkg/egressgateway/manager.go:551]: I really doubt this would cause memory issues, but if it's a major concern I could potentially try to memory profile the manager.

File: pkg/maps/egressmap/policy.go (hunk 1 above)
Comment (markpash) [reply pkg/maps/egressmap/policy.go:180]: Yeah I'm pretty sure we're ok on that front.
